llvm/unittests/IR/ConstantRangeTest.cpp

```diff
@@ -2720,6 +2720,22 @@ TEST_F(ConstantRangeTest, binaryAnd) {
   EXPECT_EQ(R16_32.binaryAnd(R0_99), R0_32);
   EXPECT_EQ(R0_99.binaryAnd(R16_32), R0_32);
 
+  // 'And' with leading bits are masked (with common leading bits stripped)
```

Comment:
What happens with signed ranges? Add a test?

Reply:
'And' op is tricky, since it operates regardless of the signedness and all ranges can be treated as unsigned. E.g., a range (with bitwidth=8) `[1, -10)` can be treated as `[1, 246)` because of the nature of 2's complement; similarly, range `[-3, -1)` can be treated as `[253, 255)`; and results won't be affected. And I think it's how `ConstantRange` designed.

I'll add a test anyway.

Reply:
Oh, if you're asking wrapped set, e.g., `[254, 1)` (or `[-2, 1)`), the algorithm bails out early at [here](https://github.com/llvm/llvm-project/commit/9c49208195b25c6cf5bdbd0f6d85ed5f8348812d#diff-b4f991f8e64d4fd23608e9ac00520c91429ad1dca8ea91f8be3a2922ea465dbdR1553). In such case, 0 is always in the range.

Reply:
Test added.